csi/src/findmnt.rs

```
@@ -0,0 +1,223 @@
+    filter: &FindmntFilter,
+    results: &mut Vec<HashMap<String, String>>,
+) {
+    if let Some(json_array) = json_val.as_array() {
+        for jsonvalue in json_array {
+            filter_findmnt(&jsonvalue, filter, results);
+        }
+    }
+    if let Some(json_map) = json_val.as_object() {
+        if let Some(value) = json_map.get(filter.key) {
+            if filter == value {
+                results.push(jsonmap_to_hashmap(json_map));
+            }
+        }
+        // If the object has arrays, then the assumption is that they are arrays
+        // of objects.
+        for (_, jsonvalue) in json_map {
+            if jsonvalue.is_array() {
+                filter_findmnt(jsonvalue, filter, results);
+            }
+        }
+    }
+}
+
+/// findmnt executable name.
+const FINDMNT: &str = "findmnt";
+/// findmnt arguments, we only want source, target and filesystem type fields.
+const FINDMNT_ARGS: [&str; 3] = ["-J", "-o", "SOURCE,TARGET,FSTYPE"];
+
+/// Execute the Linux utility findmnt, collect the json output,
+/// invoke the filter function and return the filtered results.
+fn findmnt(
+    params: FindmntFilter,
+) -> Result<Vec<HashMap<String, String>>, DeviceError> {
+    let output = Command::new(FINDMNT).args(&FINDMNT_ARGS).output()?;
+    if output.status.success() {
+        let json_str = String::from_utf8(output.stdout)?;
+        let json: Value = serde_json::from_str(&json_str)?;
+        let mut results: Vec<HashMap<String, String>> = Vec::new();
+        filter_findmnt(&json, &params, &mut results);
+        Ok(results)
+    } else {
+        Err(DeviceError {
+            message: String::from_utf8(output.stderr)?,
+        })
+    }
+}
+
+/// Use the Linux utility findmnt to find the name of the device mounted at a
+/// directory or block special file, if any.
+/// mount_path is the path a device is mounted on.
+pub(crate) fn findmnt_get_devicepath(
+    mount_path: &str,
+) -> Result<Option<String>, DeviceError> {
```

I think an Option here would be sufficient as apposed to Result<Option>

---

If I change to Option, then we would have to either 1). translate none to an error or 2) ignore the error
So we would lose precision in error reporting.